### ISSUE TITLE
TypeHinting none-entities (doctrine) causes MappingException to be thrown.

### ISSUE THREAD

**Issue by matt-usurp** [nelmio/alice#357]
I have been using `@Embeddable` objects with `Doctrine` for a while now and its gotten to the point where I have managed to push "fixtured development` at work...

Anyway, there is an issue whereby if you have a setter that `Alice` uses and the type-hint against a parameter is an object then it tries to run that through the `find()` method on `Doctrine`. See the attached link for the rough area the code fails.

https://github.com/nelmio/alice/blob/2.x/src/Nelmio/Alice/Util/TypeHintChecker.php#L71

In my case we have interfaces defined as the code base we are working with is meant to be de-coupled. So I end up with a magnitude of errors, one being that the interface cannot be found against the doctrine mappings and the second being that it is trying to look up my embeddable when it is not mean't too. Initially there was hope after being forwarded to this #272 but this doesn't help me here. I have looked at creating my own PR but I get stuck at detecting the object isn't meant to be persisted and then being left with a string `some-reference` where I have no means to resolve it to a fixtured object. Ideas?

```YAML
SomeEmbeddable (local):
  SomeEmbeddable.unlimited:
    schemeType: UNLIMITED
    customLimit: ~
    customRoutes: ~
```

```YAML
SomeUser:
  SomeUser.developer:
    email: developer@website.com
    password: developer
    authorisation: '@SomeEmbeddable.unlimited'
```

```PHP
/**
 * @ORM\Entity(repositoryClass="SomeUserRepository")
 * @ORM\Table(name="some_user")
 */
class SomeUser implements EntityInterface, UserInterface
{
    public function setSomeEmbeddable(SomeEmbeddableInterface $embeddable);
}
```

```
                                                                               
  [Doctrine\Common\Persistence\Mapping\MappingException]                                               
  Class 'SomeEmbeddableInterface' does not exist  

Exception trace:
 () at /vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/MappingException.php:96
 Doctrine\Common\Persistence\Mapping\MappingException::nonExistingClass() at /vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/RuntimeReflectionService.php:41
 Doctrine\Common\Persistence\Mapping\RuntimeReflectionService->getParentClasses() at /vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:281
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->getParentClasses() at /vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:311
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->loadMetadata() at /vendor/doctrine/orm/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php:78
 Doctrine\ORM\Mapping\ClassMetadataFactory->loadMetadata() at /vendor/doctrine/common/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:216
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->getMetadataFor() at /vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:379
 Doctrine\ORM\EntityManager->find() at /vendor/nelmio/alice/src/Nelmio/Alice/Persister/Doctrine.php:50
 Nelmio\Alice\Persister\Doctrine->find() at /vendor/nelmio/alice/src/Nelmio/Alice/Util/TypeHintChecker.php:75
 Nelmio\Alice\Util\TypeHintChecker->check() at /vendor/nelmio/alice/src/Nelmio/Alice/Instances/Populator/Methods/Direct.php:43
 Nelmio\Alice\Instances\Populator\Methods\Direct->set() at /vendor/nelmio/alice/src/Nelmio/Alice/Instances/Populator/Populator.php:86
 Nelmio\Alice\Instances\Populator\Populator->populate() at /vendor/nelmio/alice/src/Nelmio/Alice/Fixtures/Loader.php:302
 Nelmio\Alice\Fixtures\Loader->populateObjects() at /vendor/nelmio/alice/src/Nelmio/Alice/Fixtures/Loader.php:142
 Nelmio\Alice\Fixtures\Loader->load() at /vendor/hautelook/alice-bundle/src/Alice/DataFixtures/Fixtures/Loader.php:50
 Hautelook\AliceBundle\Alice\DataFixtures\Fixtures\Loader->load() at /vendor/hautelook/alice-bundle/src/Alice/DataFixtures/Loader.php:194
 Hautelook\AliceBundle\Alice\DataFixtures\Loader->tryToLoadFiles() at /vendor/hautelook/alice-bundle/src/Alice/DataFixtures/Loader.php:90
 Hautelook\AliceBundle\Alice\DataFixtures\Loader->load() at /vendor/hautelook/alice-bundle/src/Doctrine/DataFixtures/Executor/ExecutorTrait.php:41
 Hautelook\AliceBundle\Doctrine\DataFixtures\Executor\ORMExecutor->Hautelook\AliceBundle\Doctrine\DataFixtures\Executor\{closure}() at n/a:n/a
 call_user_func() at /vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php:233
 Doctrine\ORM\EntityManager->transactional() at /vendor/hautelook/alice-bundle/src/Doctrine/DataFixtures/Executor/ExecutorTrait.php:45
 Hautelook\AliceBundle\Doctrine\DataFixtures\Executor\ORMExecutor->executeExecutor() at /vendor/hautelook/alice-bundle/src/Doctrine/DataFixtures/Executor/ORMExecutor.php:72
 Hautelook\AliceBundle\Doctrine\DataFixtures\Executor\ORMExecutor->execute() at /vendor/hautelook/alice-bundle/src/Doctrine/DataFixtures/Executor/FixturesExecutor.php:44
 Hautelook\AliceBundle\Doctrine\DataFixtures\Executor\FixturesExecutor->execute() at /vendor/hautelook/alice-bundle/src/Doctrine/Command/LoadDataFixturesCommand.php:221
 Hautelook\AliceBundle\Doctrine\Command\LoadDataFixturesCommand->execute() at /vendor/symfony/symfony/src/Symfony/Component/Console/Command/Command.php:256
 Symfony\Component\Console\Command\Command->run() at /vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:804
 Symfony\Component\Console\Application->doRunCommand() at /vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:186
 Symfony\Component\Console\Application->doRun() at /vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:86
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:117
 Symfony\Component\Console\Application->run() at /bin/console:31
```

I made this issue initially against hautelook/AliceBundle#217 but I didn't realise the doctrine implementation was in this code base (partly expected it to be in the bundle) before making the issue.

**Comment by theofidry** [comment]
Thanks for the detailed bug report @matt-usurp :)

Just to make one thing clear: does changing the typehint fix the issue? I.e. the embeddables are not persisted as expected?

**Comment by matt-usurp** [comment]
Removing all TypeHints does indeed fix this issue.

**Comment by theofidry** [comment]
I'll try to check that in details in the coming week/weekend, but I can't guarantee anything. I would hate the idea of it preventing you to use interfaces, it's something that must be fixed.

**Comment by matt-usurp** [comment]
When things are less hectic at work I will stop raising issues and actually start fixing some.

**Comment by theofidry** [comment]
Ok I see where the issue comes from now. In v2, it is assumed that all your objects are getting persisted, they will be referenced by the persister (here the Doctrine manager) and leave that `find()` to the persister.

Now there is two exceptions for that:

- Embeddables objects (marked as embeddables, supported only since 2.2.0)
- local objects (your case)

One way to solve it IMO is to create a new persister for doctrine, decorating the current one, which would keep references of non-persisted objects and implement the `find()` for those ones.

**Comment by theofidry** [comment]
As shown in #427, this has been resolved in 2.2.0 :)

**Comment by matt-usurp** [comment]
Thanks for the hard work!
